roadmap/implementers-guide/src/node/disputes/dispute-coordinator.md

```diff
@@ -23,33 +28,156 @@ The meta information that we track per-candidate is defined as the `CandidateVot
+	/// that is not stable. If a new fork appears after the fact, we would start ordering the same
+	/// candidate differently, which would result in the same candidate getting queued twice.
+	relay_parent_block_number: BlockNumber,
+	/// By adding the `CandidateHash`, we can guarantee a unique ordering across candidates.
+	candidate_hash: CandidateHash,
+}
+```
+
+It also implements `PartialEq`, `Eq`, `PartialOrd` and `Ord` traits enabling comparison operations
+with the comparators.
 
-struct RecentDisputes {
-    // sorted by session index and then by candidate hash.
-    disputed: Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+`Comparator` is used inside `Participation` module as a key for saving `ParticipationRequest`. It
+provides the ordering required to process the most important requests first (check the next section
+for details).
+
+### Participation
+This module keeps track of the disputes that the node participates in. At most there are
+`MAX_PARALLEL_PARTICIPATIONS` parallel participations in the subsystem. The internal state of the
+module is:
+
+```rust
+pub struct Participation {
+	/// Participations currently being processed.
+	running_participations: HashSet<CandidateHash>,
+	/// Priority and best effort queues.
+	queue: Queues,
+	/// Sender to be passed to worker tasks.
+	worker_sender: WorkerMessageSender,
+	/// Some recent block for retrieving validation code from chain.
+	recent_block: Option<(BlockNumber, Hash)>,
 }
 ```
+New candidate is added for processing by calling `fn queue_participation()`. The function either
+starts processing immediately or queues the request. `Participation` uses another internal module
+`Queues` which provides prioritisation of the disputes. It guarantees that important disputes will
+be processed first. The actual decision how important is a given dispute is performed by the
+`ordering` module.
+
+The actual participation is performed by `fn participate()`. First it sends
+`AvailabilityRecoveryMessage::RecoverAvailableData` to obtain data from the validators. Then gets
+the validation code and stores `AvailableData` with `AvailabilityStoreMessage::StoreAvailableData`
+message.  Finally Participation module performs the actual validation and sends the result as
+`WorkerMessage` to the main subsystem (`DisputeCoordinatorSubsystem`). `Participation` generates
+messages which `DisputeCoordinatorSubsystem` consumes. You can find more information how these
+events are processed in the next section.
+
+### SpamSlots
+
+`struct SpamSlots` aims to protect the validator from malicious peers generating erroneous disputes
+with the purpose of overloading the validator with unnecessary work.
+
+How the spam protection works? Each peer validator has got a spam slot for unconfirmed disputes with
+fixed size (`MAX_SPAM_VOTES`). Each unconfirmed dispute is added to one such slot until all slots
+for the given validator are filled up. At this point all new unconfirmed disputes are ignored.
```

Comment:
statements from this validator for unconfirmed disputes are ignored .... would be a bit more precise.